contracts/Core/StakeManager.sol

```diff
@@ -173,44 +176,65 @@ contract StakeManager is Initializable, StakeStorage, StateManager, Pause, Stake
     /// @param stakerId The Id of staker associated with sRZR which user want to unstake
     /// @param sAmount The Amount in sRZR
     function unstake(uint32 stakerId, uint256 sAmount) external initialized whenNotPaused {
+        uint32 epoch = _getEpoch(epochLength);
+        require(sAmount > 0, "Non-Positive Amount");
+        require(stakerId != 0, "staker.id = 0");
+        require(stakers[stakerId].stake > 0, "Nonpositive stake");
+        require(locks[msg.sender][stakers[stakerId].tokenAddress][LockType.Unstake].amount == 0, "Existing Unstake Lock");
+
+        Structs.Staker storage staker = stakers[stakerId];
+        IStakedToken sToken = IStakedToken(staker.tokenAddress);
+
+        require(sToken.balanceOf(msg.sender) >= sAmount, "Invalid Amount");
+
+        locks[msg.sender][staker.tokenAddress][LockType.Unstake] = Structs.Lock(sAmount, 0, epoch + unstakeLockPeriod);
+        emit Unstaked(msg.sender, epoch, stakerId, sAmount, staker.stake, block.timestamp);
+
+        require(sToken.transferFrom(msg.sender, address(this), sAmount), "sToken transfer failed");
+    }
+
+    function initiateWithdraw(uint32 stakerId) external initialized whenNotPaused {
         State currentState = _getState(epochLength);
```

Comment:
require first.

Reply:
in this case, i think we should be checking for state first before going to other require statements. If we aren't in the right state, we should not go ahead with further checks.

Reply:
hmm 
though its small change
I think we should always verify input params first 
if staker id is 0 then state shouldn't be checked 
reason being we are calculating state, that's not input param 
so needlessly cal and checking that when input params are invalid, is not good approach imo

Reply:
We have actually always checked for the state first and then for input params in all our functions. It's just that in other functions, state checks are being done with the help of a modifier and here, we aren't using that since we have got check for 2 states.